### ISSUE TITLE
Notification of potential duplicates

### ISSUE THREAD

**Issue by sadatmalik** [talentbeyondboundaries/tbbtalentv2#1636]
From Zeynep:

Could we enable functionality that flags potential duplicates to admin users when multiple candidates share the same first name, last name, and birthdate? For example, an alert on a candidate’s profile could say: _“Candidate 123’s profile appears similar to Candidate 456. Please verify if they are duplicates.”_

I understand these fields are already used in periodic database checks to identify and remove duplicates. However, displaying this information to admins in real time would allow them to focus on the most current profile and help delete any additional profiles. This feature would also be especially useful in countries with multiple partners, where it would help admins identify if a candidate has registered through more than one partner.

[Slack Message](https://refugeejobsmarket.slack.com/archives/C42S4AJ73/p1730913744942839)

**Comment by samschlicht** [comment]
Sadat has already done a version of the SQL for this: 
```
WITH Duplicates AS (
    SELECT
        c.id AS candidate_id,
        u.first_name,
        u.last_name,
        c.dob,
        c.status,
        c.gender,
        u.email,
        c.phone,
        c.whatsapp,
        ed.name AS education_levl,
        co.name AS country,
        na.name AS nationality,
        c.created_date AS registration_date,
        c.updated_date AS last_updated_date,
        COUNT(*) OVER (PARTITION BY u.first_name, u.last_name, c.dob) AS dup_count
    FROM candidate c
        INNER JOIN users u ON c.user_id = u.id
        INNER JOIN country co on c.country_id = co.id
        INNER JOIN country na on c.nationality_id = na.id
        INNER JOIN education_level ed on c.max_education_level_id = ed.id
    WHERE c.status IN ('active', 'unreachable', 'incomplete', 'pending')
)
SELECT
    candidate_id,
    first_name,
    last_name,
    dob,
    status,
    gender,
    email,
    phone,
    whatsapp,
    education_levl,
    country,
    nationality,
    registration_date,
    last_updated_date
FROM Duplicates
WHERE dup_count > 1
ORDER BY first_name, last_name, status;
```

**Comment by samschlicht** [comment]
My thinking is the simplest way of adding this to candidate profiles is to leverage background batch processing and lists — e.g., once a week, candidates are added to a 'Potential duplicates' list when turned up by this query.

First thoughts on this approach, two big challenges:

- Displaying potential duplicates together (otherwise it's less useful for admins, who should be able to do this work at a glance).
  - Could potentially make a separate list for each set of potential duplicates?
  - I'll take a look at some results: if the results create some kind of order, I can replicate that in the list view.
- Removing candidates from list once the duplicate has been deleted or they've been found not to be duplicates
  - Could be handled by the background batch processing — i.e. also comb the existing lists and delete any entries no longer in the results of the duplicates query.
  - Is it too much to ask that admins remove from the list any candidates they find not to be duplicates? Perhaps not — it's just one or two clicks, depending whether they're on the candidate profile or the list view.  
 
I'll take a closer look at what exactly the query is serving up.

**Comment by samschlicht** [comment]
Just a note that the query is serving up a lot of results where name and dob are null — that might be useful for a broader scope data cleanup, but I want to focus exclusively here on potential duplicates. So I've adjusted the base query thus:
```
WITH Duplicates AS (
    SELECT
        c.id AS candidate_id,
        u.first_name,
        u.last_name,
        c.dob,
        c.status,
        c.gender,
        u.email,
        c.phone,
        c.whatsapp,
        ed.name AS education_levl,
        co.name AS country,
        na.name AS nationality,
        c.created_date AS registration_date,
        c.updated_date AS last_updated_date,
        COUNT(*) OVER (PARTITION BY u.first_name, u.last_name, c.dob) AS dup_count
    FROM candidate c
             INNER JOIN users u ON c.user_id = u.id
             INNER JOIN country co on c.country_id = co.id
             INNER JOIN country na on c.nationality_id = na.id
             INNER JOIN education_level ed on c.max_education_level_id = ed.id
    WHERE c.status IN ('active', 'unreachable', 'incomplete', 'pending')
      **AND u.first_name IS NOT NULL**
      **AND u.last_name IS NOT NULL**
      **AND c.dob IS NOT NULL**
)
SELECT
    candidate_id,
    first_name,
    last_name,
    dob,
    status,
    gender,
    email,
    phone,
    whatsapp,
    education_levl,
    country,
    nationality,
    registration_date,
    last_updated_date
FROM Duplicates
WHERE dup_count > 1
ORDER BY first_name, last_name, status;
```

**Comment by samschlicht** [comment]
There are too many potential duplicates to use lists, so my next proposal is thus:

- Add a boolean field to candidates, 'potentialDuplicate'
- Display this as a candidate row icon
- When clicked, modal opens with identified potential duplicates 
  - Could leverage ES here as for Name/Number quick search, except by DOB and first/full name - I don't want to be making join tables of duplicate groups, it's unnecessarily complicated and we can quickly serve up results when needed instead.
  - There are certainly ways that we could manage the candidate status changes from this modal, but it would be quite a lot of work for a feature that isn't super-duper important — my suggestion is to deliver a MVP along the lines of Zeynep's suggestion: the modal just serves up the potential duplicate candidates and requests the user to change status to deleted on the less-used duplicates, at the next weekly duplicate check, the candidate will no longer bear this icon if the issue has been resolved. This is plenty for someone to do this as a semi-regular task.
- NB: included in batch processing needs to be a cleanup method — any candidates for whom the potentialDuplicate is marked true, but who is not in the newly served list of potential dupes, should be marked false. 

Just noting another thought bubble: we could make this more responsive by providing a button for the admin to indicate that they've resolved the duplication by changing the status to deleted of all but one of the highlighted candidates — triggering a limited-scope version of the processing that just checks the given candidate and marks potentialDuplicate false if indeed the issue has been resolved. (The button would need to close the modal and refresh the view, ideally.)

**Comment by samschlicht** [comment]
- [x] add potentialDuplicate boolean field to candidates
- [x] write processing method that marks it true when candidate is in duplicate query results
- [x] add to it a method that cleans up using the previous results (see above for methodology)
- [x] add an icon to front end candidate results (show-candidates and candidate profile)
- [x] add a modal that identifies the potential duplicate candidates and tells the admin how to handle it and how it will be handled by the regular processing
- [x] optional extra: add something like a 'Resolved' button as described above

**Comment by samschlicht** [comment]
Here's a sudo code sketch of what I'm proposing (CandidateServiceImpl). This will need a @Transactional annotation — I would prefer to rollback an incomplete set of DB changes and go again if necessary.

<img width="751" src="https://api.zenhub.com/attachedFiles/eyJfcmFpbHMiOnsibWVzc2FnZSI6IkJBaHBBNUQ4Qnc9PSIsImV4cCI6bnVsbCwicHVyIjoiYmxvYl9pZCJ9fQ==--a70ad429274c2164215c7cdef868ff690b33e13c/Screenshot%202024-12-03%20at%201.17.05%E2%80%AFpm.png" alt="Screenshot 2024-12-03 at 1.17.05 pm.png" />

**Comment by samschlicht** [comment]
This is working now - displays on results row of candidate for whom potentialDuplicate marked true - when clicked opens a modal for links to other candidates and a resolve button for notifying when status appropriately changed, triggering refresh of results and removal of icon if server agrees. Will also have some help text for explaining process and what to do if you lack visibility to resolve yourself (notify a TC admin).
<img width="1222" src="https://api.zenhub.com/attachedFiles/eyJfcmFpbHMiOnsibWVzc2FnZSI6IkJBaHBBMno5Qnc9PSIsImV4cCI6bnVsbCwicHVyIjoiYmxvYl9pZCJ9fQ==--e273f559897119921ad1940da141199a367d6569/Screenshot%202024-12-03%20at%207.28.03%E2%80%AFpm.png" alt="Screenshot 2024-12-03 at 7.28.03 pm.png" />

**Comment by samschlicht** [comment]
Discussion from team meeting and a couple of other reminders:
- [x] add search filter
- [x] possible daily scheduling of core method
- [ ] Sam to meet w Mariam to review suitability for intern project
- [ ] get rid of individual candidate Admin Api stub but not the general duplicate check
- [ ] hot fix

**Comment by samschlicht** [comment]
Just a reminder to myself that it's by no means certain that flagged potential duplicates actually will be duplicates - the help text needs to make that clear. Without a join table I don't have an idea yet for how to mark certain seeming duplicates as safe. I think I can deliver a MVP here that doesn't include that option — I think it'll be fairly rare that all 3 of those fields match and don't point to a duplicate, we could probably live with it.

**Comment by samschlicht** [comment]
Couple of notes before I go into the wrapping-up phase of this issue (proofing, doc, testing and test-writing):

- Seeing all of this functionality is contingent on passing canViewCandidateName auth check
- It was challenging to resist the urge to go beyond MVP - with no limitations I would add these enhancements:
  - elasticsearch capability (important not to do this time because it may be a hot fix and we'd have to re-index, and also there's divergence between staging/dev and prod due to Spring upgrade) I've at least done a better job of disabling the filter.
  - modal functionality:
    - 'Not a duplicate button' with some way of excluding the candidate from future results (possibly: create a candidate note that the duplicate search checks for, make a table of excluded IDs, or add another candidate field - the challenge being that the profile could be duplicated in future! (Would probably need to use a date check to get past that.))
      - Mariam's team will need an alternative solution for this - my suggestion will be either to use an exclusion list, or reviewable saved search. 
    - 'Confirmed duplicate' button which would change the status from the modal

**Comment by samschlicht** [comment]
With a fair bit of work thinking through conflicts, Flyway issues, checking and re-checking, this is merged into staging and hot fixed into prod!

To wrap up, I'll demo this to Mariam, which I've added to my Asana board.

**Comment by samschlicht** [comment]
Final bit of business here:
Should I post a quick explainer to the TC channel so that people are forewarned that this is out there and what to do? Perhaps after running it by Zeynep and Mariam (whom I've contacted for a meeting).

And just to confirm that this is ticking away nicely on its daily schedule (2100 GMT) on prod  — with a reminder that it only picks up _new_ duplicates, so it'll be a very light load):


<img width="1293" src="https://api.zenhub.com/attachedFiles/eyJfcmFpbHMiOnsibWVzc2FnZSI6IkJBaHBBOWNVQ0E9PSIsImV4cCI6bnVsbCwicHVyIjoiYmxvYl9pZCJ9fQ==--837493ac20ce0925acc471a0b516b17016d5e8fe/Screenshot%202024-12-10%20at%2012.40.31%E2%80%AFpm.png" alt="Screenshot 2024-12-10 at 12.40.31 pm.png" />
